aws/resource_aws_ec2_traffic_mirror_session_test.go

```diff
@@ -0,0 +1,244 @@
+package aws
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+)
+
+func TestAccAWSEc2TrafficMirrorSession_basic(t *testing.T) {
+	resourceName := "aws_ec2_traffic_mirror_session.session"
+	description := "test session"
+	session := acctest.RandIntRange(1, 32766)
+	lbName := acctest.RandString(31)
+	pLen := acctest.RandIntRange(1, 255)
+	vni := acctest.RandIntRange(1, 16777216)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testAccPreCheckAWSEc2TrafficMirrorSession(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSEc2TrafficMirrorSessionDestroy,
+		Steps: []resource.TestStep{
+			//create
+			{
+				Config: testAccTrafficMirrorSessionConfig(lbName, session),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSEc2TrafficMirrorSessionExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "description", ""),
+					resource.TestCheckResourceAttr(resourceName, "packet_length", "0"),
+					resource.TestCheckResourceAttr(resourceName, "session_number", strconv.Itoa(session)),
+					resource.TestCheckNoResourceAttr(resourceName, "virtual_network_id"),
+				),
+			},
+			// update of description, packet length and VNI
+			{
+				Config: testAccTrafficMirrorSessionConfigWithOptionals(description, lbName, session, pLen, vni),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSEc2TrafficMirrorSessionExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "description", description),
+					resource.TestCheckResourceAttr(resourceName, "packet_length", strconv.Itoa(pLen)),
+					resource.TestCheckResourceAttr(resourceName, "session_number", strconv.Itoa(session)),
+					resource.TestCheckResourceAttr(resourceName, "virtual_network_id", strconv.Itoa(vni)),
+				),
+			},
+			// removal of description, packet length and VNI
+			{
+				Config: testAccTrafficMirrorSessionConfig(lbName, session),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSEc2TrafficMirrorSessionExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "description", ""),
+					resource.TestCheckResourceAttr(resourceName, "packet_length", "0"),
+					resource.TestCheckResourceAttr(resourceName, "session_number", strconv.Itoa(session)),
+					resource.TestCheckResourceAttr(resourceName, "virtual_network_id", "0"),
+				),
+			},
+			// import test without VNI
+			{
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"virtual_network_id"},
+			},
```

Comment:
Now that `virtual_network_id` is always set in the Terraform state, the following allows this test to pass:

```go
			//create
			{
				Config: testAccTrafficMirrorSessionConfig(lbName, session),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSEc2TrafficMirrorSessionExists(resourceName),
					resource.TestCheckResourceAttr(resourceName, "description", ""),
					resource.TestCheckResourceAttr(resourceName, "packet_length", "0"),
					resource.TestCheckResourceAttr(resourceName, "session_number", strconv.Itoa(session)),
					resource.TestMatchResourceAttr(resourceName, "virtual_network_id", regexp.MustCompile(`\d+`)),
				),
			},
			// update of description, packet length and VNI
			{
				Config: testAccTrafficMirrorSessionConfigWithOptionals(description, lbName, session, pLen, vni),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSEc2TrafficMirrorSessionExists(resourceName),
					resource.TestCheckResourceAttr(resourceName, "description", description),
					resource.TestCheckResourceAttr(resourceName, "packet_length", strconv.Itoa(pLen)),
					resource.TestCheckResourceAttr(resourceName, "session_number", strconv.Itoa(session)),
					resource.TestCheckResourceAttr(resourceName, "virtual_network_id", strconv.Itoa(vni)),
				),
			},
			// removal of description, packet length and VNI
			{
				Config: testAccTrafficMirrorSessionConfig(lbName, session),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSEc2TrafficMirrorSessionExists(resourceName),
					resource.TestCheckResourceAttr(resourceName, "description", ""),
					resource.TestCheckResourceAttr(resourceName, "packet_length", "0"),
					resource.TestCheckResourceAttr(resourceName, "session_number", strconv.Itoa(session)),
					resource.TestMatchResourceAttr(resourceName, "virtual_network_id", regexp.MustCompile(`\d+`)),
				),
			},
			// import test without VNI
			{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateVerify: true,
			},
```

Will handle on merge. 👍 

```
--- PASS: TestAccAWSEc2TrafficMirrorSession_basic (282.32s)
```